### ISSUE TITLE
TPU multi-processing extremely slow in comparison to GPUs

### ISSUE THREAD

**Issue by ucals** [pytorch/xla#2383]
## ❓ Questions and Help
The code below tries to train an AutoEncoder in Kaggle notebooks. It takes forever to run. I follow the [example notebooks](https://www.kaggle.com/tanlikesmath/the-ultimate-pytorch-tpu-tutorial-jigsaw-xlm-r) to the letter, and don't know what might be wrong. Kaggle's monitor shows TPU MX at constant 0% and idle time --, which indicates something is very wrong:

- Failing TPU code: https://www.kaggle.com/carlossouza/osic-autoencoder-training-on-tpus

- Working GPU code (for comparison): https://www.kaggle.com/carlossouza/osic-autoencoder-training

I've read all example notebooks and tutorials from Kaggle, and the troubleshooting documentation, but still can't figure out what's wrong, very frustrating. Please help!

Thanks

**Comment by zcain117** [comment]
Sorry to hear this isn't working well yet.

I wanted to verify: the TPU is not crashing and seems to be utilized, but the speed is slow. Does that seem right?

It looks like your model is working with pretty large examples. What is the rough size of these input examples?

2 things I wanted to mention:
1. Kaggle uses a very small VM to supply data to the TPU. When working with large input data (for example when trying classification on Imagenet) we see that the data supply is the bottleneck. Normally we would just use a more powerful VM but that's not currently an option in Kaggle as far as I know. This may or may not be your issue, I just wanted to bring it up.
2. TPU training can sometimes be slower than GPUs since it will be compiling an XLA graph for the first few steps. It's common for us to see TPU perform much slower for the first few steps or maybe the first epoch, and then once the compilations have stabilized it's typically faster. 

One thing to help debugging is right now the metrics reports are kind of confusing in your output. Could you switch to `xm.master_print(met.metrics_report())` instead of `print(met.metrics_report())`?

**Comment by zcain117** [comment]
Hey @ucals , we talked this over in my team and another idea that came forward was dynamic shapes. If your inputs have different tensor sizes, then it will cause the XLA compiler to compile a new graph for every single example, which makes training prohibitively slow. We generally fix this by forcing inputs to be the same shape (or one of a small number of shapes, e.g. 3) by padding the inputs.

Does this seem like it might apply in your case?  Or would all your inputs be the same size?

**Comment by tmabraham** [comment]
@zcain117 For what it's worth, I played around with the kernel a bit, and it seems to be that the optimizer step is actually the slowest part of the train loop.

**Comment by ucals** [comment]
Hi @zcain117 , thanks for your support! I updated the notebook, using `xm.master_print(met.metrics_report())` as you asked. 

Regarding the input, they fixed size: N x 1 x 40 x 256 x 256 (N is the batch size). They are large, though: 3D images, shaped 1 channel, 40 slices, and height & width of 256.

The `drop_last=True` in DataLoader creation ensures all the batch sizes are the same, even the last one.

**Comment by tmabraham** [comment]
@ucals It looks like it's using the TPU, but there's some bottleneck somewhere. Note that it checks TPU usage every 10 seconds so it isn't very accurate, and while there might be usage, it's likely really really low.

**Comment by jysohn23** [comment]
I forked your notebook and collected the metrics as mentioned in the TROUBLESHOOTING guide (`XLA_METRICS_FILE` env var). I took another look at the metrics this afternoon and actually it looks like the graphs are static from and it's not a dynamic input/shape issue as far as I can tell. Instead the following metrics look interesting:

Compile count is constant at 4 between steps 3 and 4. 
```
[MetricsData; step=4]
...
Metric: ExecuteTime
  TotalSamples: 5
  Accumulator: 39m09s610ms635.881us
  ValueRate: 955ms553.549us / second
  Rate: 0.00203217 / second
  Percentiles: 1%=005ms203.741us; 5%=005ms203.741us; 10%=005ms203.741us; 20%=06m38s718ms062.464us; 50%=11m10s270ms600.356us; 80%=11m10s338ms178.342us; 90%=11m10s338ms178.342us; 95%=11m10s338ms178.342us; 99%=11m10s338ms178.342us
...
Metric: OutboundData
  TotalSamples: 35
  Accumulator: 80.15MB
  ValueRate: 33.36KB / second
  Rate: 0.0142246 / second
  Percentiles: 1%=2.00B; 5%=2.00B; 10%=2.00B; 20%=2.00B; 50%=32.00B; 80%=1.95KB; 90%=27.00KB; 95%=48.00KB; 99%=80.00MB
...
Metric: ReleaseDataHandlesTime
  TotalSamples: 5
  Accumulator: 11m10s309ms214.384us
  ValueRate: 272ms436.107us / second
  Rate: 0.00203217 / second
  Percentiles: 1%=001ms332.415us; 5%=001ms332.415us; 10%=001ms332.415us; 20%=001ms445.435us; 50%=002ms992.417us; 80%=11m10s302ms960.530us; 90%=11m10s302ms960.530us; 95%=11m10s302ms960.530us; 99%=11m10s302ms960.530us
...
Metric: TransferToServerTime
  TotalSamples: 35
  Accumulator: 34m31s274ms169.674us
  ValueRate: 817ms416.376us / second
  Rate: 0.0142246 / second
  Percentiles: 1%=785.040us; 5%=840.212us; 10%=964.580us; 20%=001ms212.495us; 50%=004ms655.974us; 80%=006ms283.527us; 90%=411ms308.930us; 95%=11m10s239ms373.085us; 99%=11m10s285ms351.130us
...
Metric: XrtExecute
  TotalSamples: 43
  Accumulator: 05h13m10s798ms604.791us
  Mean: 07m17s972ms037.321us
  StdDev: 05m45s739ms255.568us
  Rate: 0.0164484 / second
  Percentiles: 25%=002ms422.511us; 50%=11m10s282ms611.132us; 80%=11m10s318ms898.775us; 90%=11m10s429ms079.627us; 95%=11m10s431ms811.287us; 99%=11m10s432ms980.554us
```

We don't really expect the execute to take this long (p50 step time looks like its 11+ mins, which is absurdly  long). Similarly for the `TransferToServerTime` the p95+ looks pretty high. I'll try to dump the graphs when I get a chance.

**Comment by jysohn23** [comment]
Attached are the graphs and metrics I've collected from a little more than an epoch.


[xla_metrics.0.txt](https://github.com/pytorch/xla/files/5007963/xla_metrics.0.txt)
[graph.6.txt](https://github.com/pytorch/xla/files/5007964/graph.6.txt)
[graph.5.txt](https://github.com/pytorch/xla/files/5007966/graph.5.txt)
[graph.4.txt](https://github.com/pytorch/xla/files/5007967/graph.4.txt)
[graph.3.txt](https://github.com/pytorch/xla/files/5007968/graph.3.txt)
[graph.2.txt](https://github.com/pytorch/xla/files/5007969/graph.2.txt)
[graph.1.txt](https://github.com/pytorch/xla/files/5007970/graph.1.txt)
[graph.0.txt](https://github.com/pytorch/xla/files/5007971/graph.0.txt)

**Comment by jysohn23** [comment]
@ucals Just a quick update that we're filing an internal bug for our XLA team to look at as the actual device step time on TPU looks like its 10min+. 3D convolutional models haven't been tested thus far with our PyTorch/XLA stack, so we hadn't gone through this code path. We'll keep you posted. For future reference, please file bugs with the output of https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information as that helps us debug and resolve issues quicker.

**Comment by ucals** [comment]
Thanks @jysohn23 !

**Comment by tmabraham** [comment]
@jysohn23 Any update on this error? Just curious...

**Comment by jysohn23** [comment]
No, unfortunately we've been pretty busy with the upcoming launch, but we do have an internal bug open with an ongoing investigation.

**Comment by tmabraham** [comment]
@jysohn23 This issue was marked stale. Any update on it?

**Comment by jysohn23** [comment]
@JackCaoG Do you happen know the followup on the internal bug?

**Comment by JackCaoG** [comment]
There is a potential optimization proposed by Blake, I haven't implement it yet. Will try to get to it when I have some circles.

**Comment by tmabraham** [comment]
@rishishounak your notebook is private so no one will be able to help you...

FYI, from the error message I see, you are not even using PyTorch you are using Keras. If so, your problem is not at all related to the issue here, or even this repository.

**Comment by rishishounak** [comment]
ok thanks

**Comment by SwapnanilMukherjee** [comment]
I am still facing the exact same issue as @tmabraham. Tried fine-tuning a VGG-16 on RGB 224x224 images but extremely slow. Batch size = 128, TPU MX=0 and Idle Time = --
I tried using a TPU with PyTorch on Colab as well, but same issue.

**Comment by JackCaoG** [comment]
@SwapnanilMukherjee can you open a new github issue? This issue is too general. If you have a chance, please try steps in https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md to do some initial analysis.
